File: backend/hilbert-main/src/main/java/com/hilbert/features/exercise/service/ExerciseService.java
package com.hilbert.features.exercise.service;

import com.hilbert.features.exercise.dto.CreateExerciseDto;
import com.hilbert.features.exercise.dto.ExerciseAnswerDto;
import com.hilbert.features.exercise.dto.ExerciseFullDto;

public interface ExerciseService {

    ExerciseFullDto getExerciseById(Long id);
    ExerciseFullDto createExercise(CreateExerciseDto exerciseDto);
    Boolean answerExercise(ExerciseAnswerDto answerDto);
    void deleteExercise(Long id);
}

File: web-frontend/src/app/features/grammar/services/exercise.service.ts
import { Injectable } from "@angular/core";
import { environment } from "../../../../environments/environment.development";
import { Observable } from "rxjs";
import { CreateExerciseDto, ExerciseAnswerDto, ExerciseFullDto } from "../models/Exercise";
import { HttpClient } from "@angular/common/http";

@Injectable({
    providedIn: "root"
})
export class ExerciseService {
    private apiUrl = `${environment.apiUrl}/exercises`;

    constructor(private http: HttpClient) {}

    getExerciseById(id: number): Observable<ExerciseFullDto> {
        return this.http.get<ExerciseFullDto>(`${this.apiUrl}/${id}`);
    }

    createExercise(exerciseDto: CreateExerciseDto): Observable<ExerciseFullDto> {
        return this.http.post<ExerciseFullDto>(this.apiUrl, exerciseDto);
    }

    answerExercise(exerciseDto: ExerciseAnswerDto): Observable<boolean> {
        return this.http.post<boolean>(`${this.apiUrl}/answer`, exerciseDto);
    }

    deleteExercise(id: number) {
        this.http.delete(`${this.apiUrl}/${id}`);
    }
}